hudi-common/src/main/java/org/apache/hudi/metadata/SecondaryIndexRecordGenerationUtils.java

```diff
@@ -278,4 +282,12 @@ public Pair<String, String> next() {
       }
     };
   }
+
+  private static Schema getRequestedSchemaForSecondaryIndex(HoodieTableMetaClient metaClient, Schema tableSchema, String secondaryKeyField) {
+    String[] recordKeyFields = metaClient.getTableConfig().getRecordKeyFields().orElse(new String[0]);
+    String[] projectionFields = Arrays.copyOf(recordKeyFields, recordKeyFields.length + 2);
+    projectionFields[recordKeyFields.length] = RECORD_KEY_METADATA_FIELD;
+    projectionFields[recordKeyFields.length + 1] = secondaryKeyField;
+    return HoodieAvroUtils.getSchemaForFields(tableSchema, Arrays.asList(projectionFields));
```

Comment:
Can we just construct the Avro schema with the avro schema builder and the `projectionFields`? Why we need to search on the table schema then?

Reply:
Projection fields are just string representation. Can you provide more details on the approach you are suggesting @danny0405 ?

Reply:
1. get the list the `Schema.Field` though `Schema.getField()`;
2. construct a new record schema through `createRecord(String name, String doc, String namespace, boolean isError, List<Field> fields)`;
3. be caution that the secondary field key might be nested, we should just use the first-level field name.

Reply:
```java
  /**
   * Fetches schema for give field names.
   * 
   * <p>If the field name is nested, only top-level field would be projected.
   * for e.g, fields ['a', 'b.c.d'] would be projected as schema (a,b).
   */
  public static Schema projectSchema(Schema oriSchema, List<String> fields) {
    List<Schema.Field> projectedFields = fields.stream()
        .map(fieldName -> oriSchema.getField(fieldName.split("\\.")[0]))
        .filter(Objects::nonNull).collect(Collectors.toList());
    return Schema.createRecord(oriSchema.getName(), oriSchema.getDoc(), oriSchema.getNamespace(), false, projectedFields);
  }
```

Reply:
Based on offline discussion, we are in alignment with the current approach.